FusionIIIT/applications/hostel_management/models.py

```diff
@@ -0,0 +1,106 @@
+import datetime
+from django.db import models
+from django.contrib.auth.models import User
+from applications.globals.models import ExtraInfo, Staff
+from applications.academic_information.models import Student
+from applications.complaint_system.models import Caretaker
+from django.utils import timezone
+
+
+class Constants:
+    ROOM_STATUS = (
+        ('Booked', 'Booked'),
+        ('CheckedIn', 'CheckedIn'),
+        ('Available', 'Available'),
+        ('UnderMaintenance', 'UnderMaintenance'),
+        )
+
+    DAYS_OF_WEEK = (
+            (0, 'Monday'),
+            (1, 'Tuesday'),
+            (2, 'Wednesday'),
+            (3, 'Thursday'),
+            (4, 'Friday'),
+            (5, 'Saturday'),
+            (6, 'Sunday')
+        )
+
+    BOOKING_STATUS = (
+    ("Confirmed" , 'Confirmed'),
+    ("Pending" , 'Pending'),
+    ("Rejected" , 'Rejected'),
+    ("Canceled" , 'Canceled'),
+    ("CancelRequested" , 'CancelRequested'),
+    ("CheckedIn" , 'CheckedIn'),
+    ("Complete", 'Complete'),
+    ("Forward", 'Forward')
+    )
+
+class GuestRoomDetail(models.Model):
+    room_no = models.CharField(max_length=4, unique=True)
+    room_status  = models.CharField(max_length=20, choices=Constants.ROOM_STATUS, default='Available')
+
+    def __str__(self):
+        return self.room_no
+
+
+class GuestDetails(models.Model):
+    guest_phone = models.CharField(max_length=15)
+    guest_name = models.CharField(max_length=40)
+    guest_email = models.CharField(max_length=40, blank=True)
+    guest_address = models.TextField(blank=True)
+    nationality = models.CharField(max_length=20, blank=True)
+
+    def __str__(self):
+        return '{} - {}'.format(self.id, self.guest_name)
+
+
+class GuestRoomBooking(models.Model):
+    caretaker = models.ForeignKey(Caretaker, on_delete=models.CASCADE)
+    student_id = models.ForeignKey(Student, on_delete=models.CASCADE)
+    guest_id = models.ForeignKey(GuestDetails, on_delete=models.CASCADE)
+    room_id = models.ManyToManyField(GuestRoomDetail)
+    number_of_rooms =  models.IntegerField(default=1,null=True,blank=True)
```

Comment:
`number_of_rooms` seems redundant as it can be derived from `room_id` field.

Reply:
ok, I will check and resolve.

Reply:
`number_of_rooms` was the derived field, not the `room_id`. We do need the `room_id` field in this model otherwise how will it be tracked which room was allotted to whom!

Reply:
In the actual form of the guest room, there is no field like which room the student wants to book. He will fill only no of rooms that he want to book not which room he want to book. And the approval of the guest room will be done in office_module.

Reply:
Okay, I got it. So, the `number_of_rooms` here is for the field present in the guestroom booking form. But we need to add the room detail somewhere (at the time of approval and not at the time of request, as to which rooms are finally allotted to the guests).

So, how the form passes? Is it first submitted to caretaker first and then goes to warden (or DSA office), or is it directly summited to the warden or DSA office. I think the process is like it first goes to the caretaker, he finds out the available room and writes that on the form and then send it to the warden or DSA office, but I might be wrong.

Reply:
ok. now, we have added a field for the room (ManyToManyField). And it will not be visible on the form that the student will book. It will be visible to those who will assign the room to the student.

Reply:
So, I think `rooms_required` would be a better field name, as `number_of_rooms` gives an idea that these number of rooms have been assigned to them, whereas these are the number of rooms which are requested.